server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -231,7 +231,7 @@ public SegmentDescriptor apply(final PartitionChunk<Sink> chunk)
                                                       // 1) Only use caching if data is immutable
                                                       // 2) Hydrants are not the same between replicas, make sure cache is local
                                                       if (hydrantDefinitelySwapped && cache.isLocal()) {
-                                                        QueryRunner<T> cachingRunner = new CachingQueryRunner<>(
+                                                        QueryRunner<T> cachingRunner = new CloseableCachingQueryRunner<>(
```

Comment:
I think you could achieve the same effect using the following code:
```java
QueryRunner<T> runner;
// 1) Only use caching if data is immutable
// 2) Hydrants are not the same between replicas, make sure cache is local
if (hydrantDefinitelySwapped && cache.isLocal()) {
  runner = CachingQueryRunner<>(...);
} else {
  runner = factory.createRunner(segment.lhs);
}
runner = QueryRunnerHelper.makeClosingQueryRunner(runner, segment.rhs);
return new Pair<>(segment.lhs.getDataInterval(), runner);
```

Thus new class `CloseableCachingQueryRunner` is not needed.

Please also:
 - Rename `segment` to something like `segmentAndCloseable`
 - Reduce the level of nesting by extracting variables, such as `FunctionalIterable.create(specs).transform(...)` as `perSegmentRunners` variable.

Reply:
It looks a good idea and I have modified the code correspondingly.
I am not very sure about "Reduce the level of nesting". I think at least we need 2-level cycles in code. One for each sink/segment in time line and the other for each hydrant in sink. So I just change FunctionalIterable.create(specs).transform() to Iterables.transform(). Is it what you want?
The new solved has been cherry-picked into 0.15.1-incubating branch locally and it works well at our cluster.